### ISSUE TITLE
Performance issues in your project (by P3)

### ISSUE THREAD

**Issue by DLPerf** [tf-encrypted/tf-encrypted#845]
Hello! I've found a performance issue in /tf_encrypted and /examples: `batch()` should be called before `map()`, which could make your program more efficient. Here is [the tensorflow document](https://tensorflow.google.cn/guide/data_performance?hl=zh_cn#vectorized_mapping) to support it.

Detailed description is listed below:

- tf_encrypted/protocol/aby3/aby3.py: `.batch(batch_size=batch_size)`[(here)](https://github.com/tf-encrypted/tf-encrypted/blob/8b7cfb32c426e9a6f56769a1b47626bd1be03a66/tf_encrypted/protocol/aby3/aby3.py#L3367) should be called before `.map(decode)`[(here)](https://github.com/tf-encrypted/tf-encrypted/blob/8b7cfb32c426e9a6f56769a1b47626bd1be03a66/tf_encrypted/protocol/aby3/aby3.py#L3365).
- examples/securenn/conv_convert.py: `.batch(batch_size)`[(here)](https://github.com/tf-encrypted/tf-encrypted/blob/8b7cfb32c426e9a6f56769a1b47626bd1be03a66/examples/securenn/conv_convert.py#L72) should be called before `.map(decoder)`[(here)](https://github.com/tf-encrypted/tf-encrypted/blob/8b7cfb32c426e9a6f56769a1b47626bd1be03a66/examples/securenn/conv_convert.py#L69) and `.map(normalize)`[(here)](https://github.com/tf-encrypted/tf-encrypted/blob/8b7cfb32c426e9a6f56769a1b47626bd1be03a66/examples/securenn/conv_convert.py#L70).
- examples/federated-learning/run.py: `dataset = dataset.batch(50)`[(here)](https://github.com/tf-encrypted/tf-encrypted/blob/8b7cfb32c426e9a6f56769a1b47626bd1be03a66/examples/federated-learning/run.py#L91) should be called before `dataset = dataset.map(decode)`[(here)](https://github.com/tf-encrypted/tf-encrypted/blob/8b7cfb32c426e9a6f56769a1b47626bd1be03a66/examples/federated-learning/run.py#L89) and `dataset = dataset.map(normalize)`[(here)](https://github.com/tf-encrypted/tf-encrypted/blob/8b7cfb32c426e9a6f56769a1b47626bd1be03a66/examples/federated-learning/run.py#L90).
- examples/federated-learning/run.py: `dataset = dataset.batch(self.BATCH_SIZE)`[(here)](https://github.com/tf-encrypted/tf-encrypted/blob/8b7cfb32c426e9a6f56769a1b47626bd1be03a66/examples/federated-learning/run.py#L157) should be called before `dataset = dataset.map(decode)`[(here)](https://github.com/tf-encrypted/tf-encrypted/blob/8b7cfb32c426e9a6f56769a1b47626bd1be03a66/examples/federated-learning/run.py#L154) and `dataset = dataset.map(normalize)`[(here)](https://github.com/tf-encrypted/tf-encrypted/blob/8b7cfb32c426e9a6f56769a1b47626bd1be03a66/examples/federated-learning/run.py#L155).
- examples/federated-learning/convert.py: `.batch(batch_size)`[(here)](https://github.com/tf-encrypted/tf-encrypted/blob/8b7cfb32c426e9a6f56769a1b47626bd1be03a66/examples/federated-learning/convert.py#L68) should be called before `.map(decode)`[(here)](https://github.com/tf-encrypted/tf-encrypted/blob/8b7cfb32c426e9a6f56769a1b47626bd1be03a66/examples/federated-learning/convert.py#L65) and `.map(normalize)`[(here)](https://github.com/tf-encrypted/tf-encrypted/blob/8b7cfb32c426e9a6f56769a1b47626bd1be03a66/examples/federated-learning/convert.py#L66).
- examples/mnist/run.py: `dataset = dataset.batch(self.BATCH_SIZE)`[(here)](https://github.com/tf-encrypted/tf-encrypted/blob/8b7cfb32c426e9a6f56769a1b47626bd1be03a66/examples/mnist/run.py#L65) should be called before `dataset = dataset.map(decode)`[(here)](https://github.com/tf-encrypted/tf-encrypted/blob/8b7cfb32c426e9a6f56769a1b47626bd1be03a66/examples/mnist/run.py#L61), `dataset = dataset.map(normalize)`[(here)](https://github.com/tf-encrypted/tf-encrypted/blob/8b7cfb32c426e9a6f56769a1b47626bd1be03a66/examples/mnist/run.py#L62) and `dataset = dataset.map(flatten)`[(here)](https://github.com/tf-encrypted/tf-encrypted/blob/8b7cfb32c426e9a6f56769a1b47626bd1be03a66/examples/mnist/run.py#L63).
- examples/mnist/run.py: `dataset = dataset.batch(self.BATCH_SIZE)`[(here)](https://github.com/tf-encrypted/tf-encrypted/blob/8b7cfb32c426e9a6f56769a1b47626bd1be03a66/examples/mnist/run.py#L147) should be called before `dataset = dataset.map(decode)`[(here)](https://github.com/tf-encrypted/tf-encrypted/blob/8b7cfb32c426e9a6f56769a1b47626bd1be03a66/examples/mnist/run.py#L144) and `dataset = dataset.map(normalize)`[(here)](https://github.com/tf-encrypted/tf-encrypted/blob/8b7cfb32c426e9a6f56769a1b47626bd1be03a66/examples/mnist/run.py#L145).
- examples/mnist/convert.py: `.batch(batch_size)`[(here)](https://github.com/tf-encrypted/tf-encrypted/blob/8b7cfb32c426e9a6f56769a1b47626bd1be03a66/examples/mnist/convert.py#L68) should be called before `.map(decode)`[(here)](https://github.com/tf-encrypted/tf-encrypted/blob/8b7cfb32c426e9a6f56769a1b47626bd1be03a66/examples/mnist/convert.py#L65) and `.map(normalize)`[(here)](https://github.com/tf-encrypted/tf-encrypted/blob/8b7cfb32c426e9a6f56769a1b47626bd1be03a66/examples/mnist/convert.py#L66).
- examples/notebooks/keras-training/common.py: `.batch(self.batch_size)`[(here)](https://github.com/tf-encrypted/tf-encrypted/blob/8b7cfb32c426e9a6f56769a1b47626bd1be03a66/examples/notebooks/keras-training/common.py#L48) should be called before `.map(norm)`[(here)](https://github.com/tf-encrypted/tf-encrypted/blob/8b7cfb32c426e9a6f56769a1b47626bd1be03a66/examples/notebooks/keras-training/common.py#L45).
- examples/notebooks/keras-training/common.py: `.batch(self.test_set_size)`[(here)](https://github.com/tf-encrypted/tf-encrypted/blob/8b7cfb32c426e9a6f56769a1b47626bd1be03a66/examples/notebooks/keras-training/common.py#L79) should be called before `.map(norm)`[(here)](https://github.com/tf-encrypted/tf-encrypted/blob/8b7cfb32c426e9a6f56769a1b47626bd1be03a66/examples/notebooks/keras-training/common.py#L78).
- examples/logistic/common.py: `.batch(self.batch_size)`[(here)](https://github.com/tf-encrypted/tf-encrypted/blob/8b7cfb32c426e9a6f56769a1b47626bd1be03a66/examples/logistic/common.py#L118) should be called before `.map(norm)`[(here)](https://github.com/tf-encrypted/tf-encrypted/blob/8b7cfb32c426e9a6f56769a1b47626bd1be03a66/examples/logistic/common.py#L115).
- examples/logistic/common.py: `.batch(self.test_set_size)`[(here)](https://github.com/tf-encrypted/tf-encrypted/blob/8b7cfb32c426e9a6f56769a1b47626bd1be03a66/examples/logistic/common.py#L149) should be called before `.map(norm)`[(here)](https://github.com/tf-encrypted/tf-encrypted/blob/8b7cfb32c426e9a6f56769a1b47626bd1be03a66/examples/logistic/common.py#L148).
- examples/logistic/data.py: `.batch(batch_size)`[(here)](https://github.com/tf-encrypted/tf-encrypted/blob/8b7cfb32c426e9a6f56769a1b47626bd1be03a66/examples/logistic/data.py#L23) should be called before `.map(norm)`[(here)](https://github.com/tf-encrypted/tf-encrypted/blob/8b7cfb32c426e9a6f56769a1b47626bd1be03a66/examples/logistic/data.py#L20).
- examples/logistic/data.py: `.batch(batch_size)`[(here)](https://github.com/tf-encrypted/tf-encrypted/blob/8b7cfb32c426e9a6f56769a1b47626bd1be03a66/examples/logistic/data.py#L41) should be called before `.map(norm)`[(here)](https://github.com/tf-encrypted/tf-encrypted/blob/8b7cfb32c426e9a6f56769a1b47626bd1be03a66/examples/logistic/data.py#L40).

Besides, you need to check the function called in `map()`(e.g., `norm` called in `map()`) whether to be affected or not to make the changed code work properly. For example, if `norm` needs data with shape (x, y, z) as its input before fix, it would require data with shape (batch_size, x, y, z).

Looking forward to your reply. Btw, I am very glad to create a PR to fix it if you are too busy.

**Comment by zicofish** [comment]
I will close this issue for now.
